File: kotlin/src/main/kotlin/edax/Board.kt
package edax

/**
 * Othello/Reversi board representation using bitboards.
 * Each square is represented as a bit in a 64-bit integer.
 * Square 0 = A1, Square 7 = H1, Square 56 = A8, Square 63 = H8.
 */
data class Board(
    var player: ULong,
    var opponent: ULong
) {
    companion object {
        const val INIT_PLAYER: ULong = 0x0000000810000000uL   // E4, D5
        const val INIT_OPPONENT: ULong = 0x0000001008000000uL // D4, E5
        const val INIT_MOVES: ULong = 0x0000102004080000uL    // D3, C4, F5, E6

        /** Create a new board with the standard initial position. */
        fun new(): Board = Board(INIT_PLAYER, INIT_OPPONENT)

        /** Convert square index to algebraic notation. */
        fun squareToString(square: Int): String {
            val col = ('A' + (square % 8))
            val row = (square / 8) + 1
            return "$col$row"
        }
    }

    /** Pass the turn (swap player and opponent). */
    fun pass() {
        val tmp = player
        player = opponent
        opponent = tmp
    }

    /** Return the number of empty squares. */
    fun empties(): Int = 64 - (player or opponent).countOneBits()

    /** Return the number of player discs. */
    fun playerDiscCount(): Int = player.countOneBits()

    /** Return the number of opponent discs. */
    fun opponentDiscCount(): Int = opponent.countOneBits()

    /** Return the score (player discs - opponent discs). */
    fun score(): Int {
        val p = playerDiscCount()
        val o = opponentDiscCount()
        val e = empties()
        return when {
            p > o -> p - o + e
            p < o -> p - o - e
            else -> 0
        }
    }

    /**
     * Get the color at a specific square.
     * Returns: 0 = player, 1 = opponent, 2 = empty
     */
    fun getSquareColor(square: Int): Int {
        val mask = 1uL shl square
        return when {
            (player and mask) != 0uL -> 0
            (opponent and mask) != 0uL -> 1
            else -> 2
        }
    }

    /** Get all legal moves for the current player. */
    fun getMoves(): ULong = getMoves(player, opponent)

    /** Get legal moves using the parallel prefix algorithm. */
    fun getMoves(P: ULong, O: ULong): ULong {
        val mask = O and 0x7E7E7E7E7E7E7E7EuL
        var moves: ULong = 0uL
        val empty = (P or O).inv()

        // Right direction (+1)
        var flip = mask and (P shl 1)
        flip = flip or (mask and (flip shl 1))
        flip = flip or (mask and (flip shl 1))
        flip = flip or (mask and (flip shl 1))
        flip = flip or (mask and (flip shl 1))
        flip = flip or (mask and (flip shl 1))
        moves = moves or (flip shl 1)

        // Left direction (-1)
        flip = mask and (P shr 1)
        flip = flip or (mask and (flip shr 1))
        flip = flip or (mask and (flip shr 1))
        flip = flip or (mask and (flip shr 1))
        flip = flip or (mask and (flip shr 1))
        flip = flip or (mask and (flip shr 1))
        moves = moves or (flip shr 1)

        // Down direction (+8)
        flip = O and (P shl 8)
        flip = flip or (O and (flip shl 8))
        flip = flip or (O and (flip shl 8))
        flip = flip or (O and (flip shl 8))
        flip = flip or (O and (flip shl 8))
        flip = flip or (O and (flip shl 8))
        moves = moves or (flip shl 8)

        // Up direction (-8)
        flip = O and (P shr 8)
        flip = flip or (O and (flip shr 8))
        flip = flip or (O and (flip shr 8))
        flip = flip or (O and (flip shr 8))
        flip = flip or (O and (flip shr 8))
        flip = flip or (O and (flip shr 8))
        moves = moves or (flip shr 8)

        // Down-right direction (+9)
        flip = mask and (P shl 9)
        flip = flip or (mask and (flip shl 9))
        flip = flip or (mask and (flip shl 9))
        flip = flip or (mask and (flip shl 9))
        flip = flip or (mask and (flip shl 9))
        flip = flip or (mask and (flip shl 9))
        moves = moves or (flip shl 9)

        // Up-left direction (-9)
        flip = mask and (P shr 9)
        flip = flip or (mask and (flip shr 9))
        flip = flip or (mask and (flip shr 9))
        flip = flip or (mask and (flip shr 9))
        flip = flip or (mask and (flip shr 9))
        flip = flip or (mask and (flip shr 9))
        moves = moves or (flip shr 9)

        // Down-left direction (+7)
        flip = mask and (P shl 7)
        flip = flip or (mask and (flip shl 7))
        flip = flip or (mask and (flip shl 7))
        flip = flip or (mask and (flip shl 7))
        flip = flip or (mask and (flip shl 7))
        flip = flip or (mask and (flip shl 7))
        moves = moves or (flip shl 7)

        // Up-right direction (-7)
        flip = mask and (P shr 7)
        flip = flip or (mask and (flip shr 7))
        flip = flip or (mask and (flip shr 7))
        flip = flip or (mask and (flip shr 7))
        flip = flip or (mask and (flip shr 7))
        flip = flip or (mask and (flip shr 7))
        moves = moves or (flip shr 7)

        return moves and empty
    }

    /** Calculate the discs that would be flipped by placing at `square`. */
    fun flip(square: Int): ULong = flipDisc(square, player, opponent)

    /** Flip calculation implementation. */
    fun flipDisc(square: Int, P: ULong, O: ULong): ULong {
        val move = 1uL shl square
        var totalFlip: ULong = 0uL

        val directions = intArrayOf(1, -1, 8, -8, 9, -9, 7, -7)
        val masks = ulongArrayOf(
            0x7E7E7E7E7E7E7E7EuL, 0x7E7E7E7E7E7E7E7EuL,
            0xFFFFFFFFFFFFFFFFuL, 0xFFFFFFFFFFFFFFFFuL,
            0x7E7E7E7E7E7E7E7EuL, 0x7E7E7E7E7E7E7E7EuL,
            0x7E7E7E7E7E7E7E7EuL, 0x7E7E7E7E7E7E7E7EuL
        )

        for (i in 0 until 8) {
            val m = masks[i] and O
            val d = directions[i]
            var f: ULong = 0uL
            var x = move

            if (d > 0) {
                x = x shl d
                while ((x and m) != 0uL) {
                    f = f or x
                    x = x shl d
                }
                if ((x and P) != 0uL) totalFlip = totalFlip or f
            } else {
                val nd = -d
                x = x shr nd
                while ((x and m) != 0uL) {
                    f = f or x
                    x = x shr nd
                }
                if ((x and P) != 0uL) totalFlip = totalFlip or f
            }
        }

        return totalFlip
    }

    /** Play a move at the given square. Returns the flipped discs for undo. */
    fun doMove(square: Int): ULong {
        val flipped = flip(square)
        player = player or flipped or (1uL shl square)
        opponent = opponent xor flipped
        pass()
        return flipped
    }

    /** Undo a move. */
    fun undoMove(square: Int, flipped: ULong) {
        pass()
        player = player xor flipped xor (1uL shl square)
        opponent = opponent xor flipped
    }

    /** Check if the game is over (neither player can move). */
    fun isGameOver(): Boolean {
        if (getMoves() != 0uL) return false
        val passed = copy()
        passed.pass()
        return passed.getMoves() == 0uL
    }

    /** Check if the current player must pass. */
    fun isPass(): Boolean = getMoves() == 0uL

    /** Get mobility (number of legal moves). */
    fun mobility(): Int = getMoves().countOneBits()
}

File: swift/Sources/EdaxCore/Board.swift
/// Othello/Reversi board representation using bitboards.
/// Each square is represented as a bit in a 64-bit integer.
/// Square 0 = A1, Square 7 = H1, Square 56 = A8, Square 63 = H8.
public struct Board: Equatable, Hashable {
    public var player: UInt64
    public var opponent: UInt64

    /// Initial position constants
    public static let initPlayer: UInt64 = 0x0000000810000000    // E4, D5
    public static let initOpponent: UInt64 = 0x0000001008000000  // D4, E5
    public static let initMoves: UInt64 = 0x0000102004080000    // D3, C4, F5, E6

    /// Create a new board with the standard initial position.
    public init() {
        self.player = Board.initPlayer
        self.opponent = Board.initOpponent
    }

    /// Create a board with specified player and opponent bitboards.
    public init(player: UInt64, opponent: UInt64) {
        self.player = player
        self.opponent = opponent
    }

    /// Pass the turn (swap player and opponent).
    public mutating func pass() {
        swap(&player, &opponent)
    }

    /// Return the number of empty squares.
    public func empties() -> Int {
        return 64 - (player | opponent).nonzeroBitCount
    }

    /// Return the number of player discs.
    public func playerDiscCount() -> Int {
        return player.nonzeroBitCount
    }

    /// Return the number of opponent discs.
    public func opponentDiscCount() -> Int {
        return opponent.nonzeroBitCount
    }

    /// Return the score (player discs - opponent discs).
    /// Empty squares are awarded to the leader.
    public func score() -> Int {
        let p = playerDiscCount()
        let o = opponentDiscCount()
        let e = empties()
        if p > o {
            return p - o + e
        } else if p < o {
            return p - o - e
        } else {
            return 0
        }
    }

    /// Get the color at a specific square.
    /// Returns: 0 = player, 1 = opponent, 2 = empty
    public func getSquareColor(_ square: Int) -> Int {
        let mask: UInt64 = 1 << square
        if (player & mask) != 0 { return 0 }
        if (opponent & mask) != 0 { return 1 }
        return 2
    }

    /// Get all legal moves for the current player.
    public func getMoves() -> UInt64 {
        return getMoves(player: player, opponent: opponent)
    }

    /// Get legal moves using the parallel prefix algorithm.
    public func getMoves(player P: UInt64, opponent O: UInt64) -> UInt64 {
        let mask = O & 0x7E7E7E7E7E7E7E7E
        var moves: UInt64 = 0
        let empty = ~(P | O)

        // Right direction (+1)
        var flip = mask & (P << 1)
        flip |= mask & (flip << 1)
        flip |= mask & (flip << 1)
        flip |= mask & (flip << 1)
        flip |= mask & (flip << 1)
        flip |= mask & (flip << 1)
        moves |= flip << 1

        // Left direction (-1)
        flip = mask & (P >> 1)
        flip |= mask & (flip >> 1)
        flip |= mask & (flip >> 1)
        flip |= mask & (flip >> 1)
        flip |= mask & (flip >> 1)
        flip |= mask & (flip >> 1)
        moves |= flip >> 1

        // Down direction (+8)
        flip = O & (P << 8)
        flip |= O & (flip << 8)
        flip |= O & (flip << 8)
        flip |= O & (flip << 8)
        flip |= O & (flip << 8)
        flip |= O & (flip << 8)
        moves |= flip << 8

        // Up direction (-8)
        flip = O & (P >> 8)
        flip |= O & (flip >> 8)
        flip |= O & (flip >> 8)
        flip |= O & (flip >> 8)
        flip |= O & (flip >> 8)
        flip |= O & (flip >> 8)
        moves |= flip >> 8

        // Down-right direction (+9)
        flip = mask & (P << 9)
        flip |= mask & (flip << 9)
        flip |= mask & (flip << 9)
        flip |= mask & (flip << 9)
        flip |= mask & (flip << 9)
        flip |= mask & (flip << 9)
        moves |= flip << 9

        // Up-left direction (-9)
        flip = mask & (P >> 9)
        flip |= mask & (flip >> 9)
        flip |= mask & (flip >> 9)
        flip |= mask & (flip >> 9)
        flip |= mask & (flip >> 9)
        flip |= mask & (flip >> 9)
        moves |= flip >> 9

        // Down-left direction (+7)
        flip = mask & (P << 7)
        flip |= mask & (flip << 7)
        flip |= mask & (flip << 7)
        flip |= mask & (flip << 7)
        flip |= mask & (flip << 7)
        flip |= mask & (flip << 7)
        moves |= flip << 7

        // Up-right direction (-7)
        flip = mask & (P >> 7)
        flip |= mask & (flip >> 7)
        flip |= mask & (flip >> 7)
        flip |= mask & (flip >> 7)
        flip |= mask & (flip >> 7)
        flip |= mask & (flip >> 7)
        moves |= flip >> 7

        return moves & empty
    }

    /// Calculate the discs that would be flipped by placing at `square`.
    public func flip(_ square: Int) -> UInt64 {
        return flipDisc(square, player: player, opponent: opponent)
    }

    /// Flip calculation implementation.
    public func flipDisc(_ square: Int, player P: UInt64, opponent O: UInt64) -> UInt64 {
        let move: UInt64 = 1 << square
        var totalFlip: UInt64 = 0

        let directions = [1, -1, 8, -8, 9, -9, 7, -7]
        let masks: [UInt64] = [
            0x7E7E7E7E7E7E7E7E, 0x7E7E7E7E7E7E7E7E,
            0xFFFFFFFFFFFFFFFF, 0xFFFFFFFFFFFFFFFF,
            0x7E7E7E7E7E7E7E7E, 0x7E7E7E7E7E7E7E7E,
            0x7E7E7E7E7E7E7E7E, 0x7E7E7E7E7E7E7E7E
        ]

        for i in 0..<8 {
            let m = masks[i] & O
            let d = directions[i]
            var f: UInt64 = 0
            var x = move

            if d > 0 {
                x = x << d
                while (x & m) != 0 {
                    f |= x
                    x = x << d
                }
                if (x & P) != 0 { totalFlip |= f }
            } else {
                let nd = -d
                x = x >> nd
                while (x & m) != 0 {
                    f |= x
                    x = x >> nd
                }
                if (x & P) != 0 { totalFlip |= f }
            }
        }

        return totalFlip
    }

    /// Play a move at the given square.
    /// Returns the flipped discs for undo.
    @discardableResult
    public mutating func doMove(_ square: Int) -> UInt64 {
        let flipped = flip(square)
        player |= flipped | (1 << square)
        opponent ^= flipped
        swap(&player, &opponent)
        return flipped
    }

    /// Undo a move.
    public mutating func undoMove(_ square: Int, flipped: UInt64) {
        swap(&player, &opponent)
        player ^= flipped | (1 << square)
        opponent ^= flipped
    }

    /// Check if the game is over (neither player can move).
    public func isGameOver() -> Bool {
        if getMoves() != 0 { return false }
        var passed = self
        passed.pass()
        return passed.getMoves() == 0
    }

    /// Check if the current player must pass.
    public func isPass() -> Bool {
        return getMoves() == 0
    }

    /// Get mobility (number of legal moves).
    public func mobility() -> Int {
        return getMoves().nonzeroBitCount
    }

    /// Convert square index to algebraic notation.
    public static func squareToString(_ square: Int) -> String {
        let col = Character(UnicodeScalar(UInt8(65 + square % 8)))
        let row = (square / 8) + 1
        return "\(col)\(row)"
    }
}
